lib/livebook/runtime/evaluator.ex

```diff
@@ -688,20 +695,87 @@ defmodule Livebook.Runtime.Evaluator do
     {result, code_markers}
   end
 
+  # Erlang code is either statements as currently supported, or modules.
+  # In case we want to support modules - it makes sense to allow users to use
+  # includes, defines and thus we use the epp-module first - try to find out
+  # if we have a module attribute, and if so deem it a module.
+  # If no module attribute was found the previous code is called.
   defp eval(:erlang, code, binding, env) do
+    filename = tmp_filename()
+    :file.write_file(filename, code)
+
+    eval_result =
+      case :epp.parse_file(filename, [], []) do
+        {:ok, forms} ->
+          case :lists.keyfind(:module, 3, forms) do
+            {:attribute, _lineno, :module, module} ->
+              eval_erlang_module(code, module, forms, binding, env)
+
+            false ->
+              case :erl_scan.string(String.to_charlist(code), {1, 1}, [:text]) do
+                {:ok, tokens, _} ->
+                  eval_erlang_statements(code, tokens, binding, env)
+
+                {:error, {location, module, description}, _end_loc} ->
+                  process_erlang_error(env, code, location, module, description)
+              end
+          end
+
+        {:error, epp_parse_errors} ->
+          {{:error, :error, epp_parse_errors, []}, []}
+      end
+
+    # Clean up after ourselves.
+    :file.delete(filename)
+    eval_result
+  end
+
+  # Create module - tokens from string
+  # Based on: https://stackoverflow.com/questions/2160660/how-to-compile-erlang-code-loaded-into-a-string
+  # Step 1: Scan the code using the epp:parse_file erlang primitive
+  # - epp will do macro expansion etc.
+  # Step 2: Convert to forms
+  # Step 3: Extract module name
+  # Step 4: Compile and load
+  # Step 5: If compile success - register module
+
+  defp eval_erlang_module(_code, module, forms, binding, env) do
+    try do
+      case :compile.forms(forms) do
+        {:ok, _, binary_module} ->
+          {:module, module} = :code.load_binary(module, ~c"#{module}.beam", binary_module)
```

Comment:
Technically `:code.load_binary` doesn't use the filename, but if we want to be accurate, we can do this:

```suggestion
          file =
            if ebin_path = ebin_path() do
              Path.join(ebin_path, "#{module}.beam")
            else
              "#{module}.beam"
            end

          {:module, module} = :code.load_binary(module, String.to_charlist(file), binary_module)
```